File: IExpr.fs
module IExpr
open System 
//The IExpr is a type with the operators of the language
type IExpr = Cons of int 
            | Neg of IExpr
            | Abs of IExpr
            | Plus of IExpr * IExpr
            | Times of IExpr * IExpr
            | Minus of IExpr * IExpr
            | Exp of IExpr * IExpr
//pow function is used to calculate the exp
let pow x y =
    let mutable r = 1
    let mutable newY = y
    while newY > 0 do
        r <- r * x
        newY <- newY - 1
    r

//interpret recursively calls it self to interpret to expr
// the base case is when the expr matches with Cons
let rec interpret expr = 
    match expr with 
    | Cons c1 -> c1
    | Neg c1 -> -interpret c1 
    | Abs c1 ->  if (interpret c1) < 0 then (-interpret c1) else interpret c1
    | Plus (c1,c2) -> (interpret c1) + (interpret c2) 
    | Times (c1,c2) -> (interpret c1) * (interpret c2) 
    | Minus (c1,c2) -> (interpret c1) - (interpret c2) 
    | Exp (c1,c2) -> pow (interpret c1) (interpret c2) 


(*  
let cons1 = Cons (5)
let cons2 = Cons (6)
let neg = Neg(cons1)
let abs = Abs(neg)
let plus = Plus(cons1,cons2)
let times = Times(cons1,cons2)
let minus = Minus(cons1,cons2)
let exp = Exp(cons1,cons2)
printfn "cons 5: %d" (interpret cons1)
printfn "neg cons 5: %d" (interpret neg)
printfn "abs neg cons 5: %d" (interpret abs)
printfn "plus cons 5 cons 6: %d" (interpret plus)
printfn "times cons 5 cons 6: %d" (interpret times)
printfn "exp cons 5 cons 6: %d" (interpret exp)*)

File: IExpr.rb
#pow function is used to calculate the exponent
def pow(x,y)
	r = 1
	while y > 0
		r *= x
		y -= 1
	end
	return r
end
#IExpr is a class with many subclasses
#The subclasses are the operators of the language
class IExpr
	def interpret()
		puts("Wrong action")
	end

end
class Cons < IExpr
	attr_accessor :value
	def initialize(val)
		@value = val
	end
	def interpret()
		return value
	end
end
class Neg < IExpr
	attr_accessor :cons
	def initialize(cons)
		@cons = cons
	end
	def interpret()
		constant = cons.interpret()
		return (- constant)
	end
end
class Abs < IExpr 
	attr_accessor :expr
	def initialize(value)
		@expr = value
	end
	def interpret()
		exprValue = expr.interpret()
		#this checks if the input is negative or not
		return (exprValue < 0 ? -exprValue : exprValue)
	end
end

class Plus < IExpr
	attr_accessor :expr1, :expr2
	def initialize(obj1,obj2)
		@expr1 = obj1
		@expr2 = obj2
	end
	def interpret()
		return expr1.interpret() + expr2.interpret()
	end
end
class Times < IExpr
	attr_accessor :expr1, :expr2
	def initialize(obj1, obj2)
		@expr1 = obj1
		@expr2 = obj2
	end
	def interpret()
		return expr1.interpret() * expr2.interpret()
	end
end

class Minus < IExpr
	attr_accessor :expr1, :expr2
	def initialize(obj1,obj2)
		@expr1 = obj1
		@expr2 = obj2
	end
	def interpret()
		return expr1.interpret() - expr2.interpret()
	end
end

class Exp < IExpr
        attr_accessor :expr1, :expr2
        def initialize(obj1,obj2)
                @expr1 = obj1
                @expr2 = obj2
        end
        def interpret()
		r1 = expr1.interpret()
		r2 = expr2.interpret()
		return pow(r1,r2)
        end
end

=begin
	testing
	cons1 = Cons.new(5)
	cons2 = Cons.new(6)
	neg = Neg.new(cons1)
	abs = Abs.new(neg)
	plus = Plus.new(cons1, cons2)
	times = Times.new(plus, cons2)
	minus = Minus.new(plus, cons2)
	exp = Exp.new(cons1,cons2)
	puts("const: #{cons1.interpret()}")
	puts("neg: #{neg.interpret()}")
	puts("abs: #{abs.interpret()}")
	puts("plus: #{plus.interpret()}")
	puts("times: #{times.interpret()}")
	puts("minus: #{minus.interpret()}")
	puts("exp: #{exp.interpret()}")
=end

